### PR TITLE
[v7] Skip tests on a docs-only PR (#9416)

### PR DESCRIPTION
**Pull request by tcsc**
Skips long-running tests if only doc changes are detected in a PR. Uses
a Go script to compare the changes between the target branch and the
supplied PR code. If those changes are only to files deemed to be docs
(i.e. docs/**, **.md or **.mdx), then the script will finish early and return
success.

If the script detects non-doc changes it will go on to build and run the
appropriate tests.

### CODE REVIEW

File: .cloudbuild/scripts/internal/changes/changes.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+	return report, nil
+}
+
+func isDocChange(path string) bool {
+	path = strings.ToLower(path)
+	return strings.HasPrefix(path, "docs/") ||
+		strings.HasSuffix(path, ".mdx") ||
+		strings.HasSuffix(path, ".md")
+}
```
Comment (quinqu) [inline .cloudbuild/scripts/internal/changes/changes.go:81]:
```suggestion
	path = strings.ToLower(path)
	return strings.HasPrefix(path, "docs/") ||
		strings.HasSuffix(path, ".mdx") ||
		strings.HasSuffix(path, ".md") || 
		strings.HasPrefix(path, "rfd/")
}
```

File: .cloudbuild/scripts/cmd/integration-tests/main.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		log.Println("No non-docs changes detected. Skipping tests.")
+		return nil
+	}
+
+	log.Printf("Running root-only integration tests...")
+	err = runRootIntegrationTests(args.workspace, gomodcache)
+	if err != nil {
+		return trace.Wrap(err, "Root-only integration tests failed")
+	}
+
```
Comment (quinqu) [inline .cloudbuild/scripts/cmd/integration-tests/main.go:115]:
```suggestion
log.Println("Root-only integration tests passed.")
```

Comment (quinqu) [reply .cloudbuild/scripts/cmd/integration-tests/main.go:115]: Adding this extra log-line can pin down potential failures in the future.

File: .cloudbuild/scripts/cmd/integration-tests/main.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+		// want the nonroot user to be able to manipulate the artifacts
+		// created by root, so we `chown -R` the whole workspace to allow it.
+		log.Printf("Reconfiguring workspace for nonroot user")
+		err = chownR(args.workspace, nonrootUID, nonrootGID)
+		if err != nil {
+			return trace.Wrap(err, "failed reconfiguring workspace")
+		}
+	}
+
+	log.Printf("Starting etcd...")
```
Comment (quinqu) [inline .cloudbuild/scripts/cmd/integration-tests/main.go:127]: Are the `ectd` tests the only tests we run as non-root?

File: .cloudbuild/scripts/cmd/integration-tests/main.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+		return trace.Wrap(err, "failed starting etcd")
+	}
+
+	log.Printf("Running nonroot integration tests...")
+	err = runNonrootIntegrationTests(args.workspace, nonrootUID, nonrootGID, gomodcache)
+	if err != nil {
+		return trace.Wrap(err, "Nonroot integration tests failed")
+	}
+
+	log.Printf("PASS")
```
Comment (quinqu) [inline .cloudbuild/scripts/cmd/integration-tests/main.go:141]:
```suggestion
	log.Printf("Non-root integration tests passed.")
```

File: .cloudbuild/scripts/cmd/unit-tests/main.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+	flag.Parse()
+
+	if args.workspace == "" {
+		return args, trace.Errorf("workspace path must be set")
+	}
+
+	var err error
+	args.workspace, err = filepath.Abs(args.workspace)
+	if err != nil {
+		return args, trace.Wrap(err, "Unable to resole absolute path to workspace")
```
Comment (quinqu) [inline .cloudbuild/scripts/cmd/unit-tests/main.go:68]:
```suggestion
		return args, trace.Wrap(err, "Unable to resolve absolute path to workspace")
```

File: .cloudbuild/scripts/cmd/integration-tests/main.go (hunk 3 above)
Comment (tcsc) [reply .cloudbuild/scripts/cmd/integration-tests/main.go:127]: We run a heap of tests as nonroot - the `etcd` tests just happen to be among them. I tried running `etcd` as root and the tests as nonroot. Unfortunately this produces a heap of spurious error messages[1] when running the tests that I did not want confusing people in the logs. 

1. The files created by `etcd` live inside the directory searched by `go list ./...` and so on to generate the list of package names, and so making them owned by `root` produces a heap of diagnostic warnings that would pollute the build log and just confuse people when they are trying to work out why their build failed. 

...Perhaps this should be a comment in the program :-)

**Review by quinqu — APPROVED** [review]

**Review by russjones — APPROVED** [review]

**Pull request merged**